### PR TITLE
feat: tighten submission validation

### PR DESCRIPTION
**Pull request by mantariksh**
## Problem

Field validation has various related issues (described below) which allow responses with invalid shapes to slip past. This means that code after the validation step incorrectly assumes that the responses have a fixed shape, which can lead to null pointer exceptions.

## Solution

**Joi validation on field types**
The Joi validation on the submission endpoints enforces that `fieldType` is a string, but not that it matches one of our valid field types. Subsequent code simply assumes that the field type is a valid member of our `BasicField` enum.

The solution was to add the list of valid field types to our Joi validation on the `fieldType` key in each response.

**Joi validation on admin preview endpoint**
The preview submissions endpoint did not have any Joi validation, even though it called the same middleware functions as the public submissions endpoint. Hence identical validation was added at the preview endpoint.

**Response type guards**
The old `src/types/response/guards.ts` had a `isProcessedSingleAnswerResponse` function which took an argument of type `any` and checked if it was a valid `ProcessedSingleAnswerResponse`. I thought this function could be improved in the following ways:
1. Since it is currently only called with an argument of type `ProcessedFieldResponse`, its argument can be scoped to that type.
2. It should check that the field type is actually a single answer field type, so that for e.g. you can't submit a response of fieldType `checkbox` but with `answer` instead of `answerArray`.

To this end, I made the following changes:
- Changed the argument of `isProcessedSingleAnswerResponse` to be of type `ProcessedFieldResponse`, and removed the helper function `isProcessedFieldResponse` which checks for the shape of `ProcessedFieldResponse`. Together, the validation steps proceed as follows:
    1. Joi validation ensures that each response has a minimum of `_id`, `question` and `fieldType`, which is the minimum for a `FieldResponse`.
    2. The `getProcessedResponses` function turns the `FieldResponse` into a `ProcessedFieldResponse` by adding on the optional `isVisible` and `isUserVerified` booleans.
    3. The `isProcessedSingleAnswerResponse` type guard ensures that the response contains a string `answer` and that the field type is a single-answer field type.
- Added an `answerArray` boolean to `shared/resources/basic` to differentiate between fields with answer vs answerArray.
- Restricted the types in `src/types/response/index.ts` to accept only valid field types for each type of response.

**Control flow in `validateField`**
The old control flow in the `validateField` function worked as follows:
1. If the response has the correct shape for a single answer response, perform either FP or class-based validation.
2. Else, default to class-based validation.

The problem here is that we are defaulting to class-based validation for table/checkbox fields, without checking that the answers have the correct shape. This can lead to null-pointer exceptions if, say, someone submits a single-answer field with `answerArray` instead of `answer`.

The solution was to add type guards for `ProcessedCheckboxResponse` and `ProcessedTableResponse`, and explicitly check for the correct shape before defaulting to class-based validation. If none of the valid shapes are matched, an error is returned.

**SPCP fake field types**
For SPCP forms, the NRIC or UEN/UID are appended to the parsed responses with the same shape as regular responses so that they can be rendered in the email or the data tab. Previously, they were given fake `authenticationSp` and `authenticationCp` field types, which no longer worked after the Response field types were restricted. Since the field types are not used anyway, these were changed to `BasicField.Nric` (for NRIC and UID) and `BasicField.ShortText` (for UEN).

**Jasmine tests**
The tighter validation resulted in some Jasmine tests breaking because our tests are somehow checking that validation passes when a `table` field is submitted with `answer` instead of `answerArray`. These tests were updated.

**Jest tests**
The tests for `submission.service` failed to compile after I tightened the types. I found these tests quite difficult to work with due to the complexity of the input data, so I extracted a bunch of helper functions to generate fake form fields and responses, and simplified the input data of the tests.

## Tests
- [ ] Submit an email mode form with all form fields in preview mode.
- [ ] Submit a storage mode form with all form fields in preview mode.
- [ ] Create a Singpass email mode form with an autoreply email field. Submit the form and check that the NRIC is rendered correctly in both the admin response email and the autoreply email.
- [ ] Create a Corppass email mode form with an autoreply email field. Submit the form and check that the UEN and UID are rendered correctly in both the admin response email and the autoreply email.
- [ ] Create a Singpass storage mode form. Submit the form and check that the NRIC is rendered correctly in the data tab and the exported CSV. Add a form field and submit the form again. Check that the exported CSV has all the NRICs in the same column.
- [ ] Programmatically submit a table field with no `answerArray`, but with `answer` as an empty string. Check that you get a 400 and not a 500 response.
- [ ] Programmatically submit a table field with no `answerArray`, but with `answer` as a non-empty string. Check that you get a 400 and not a 500 response.
- [ ] Programmatically submit a short text field with no `answer`, but with `answerArray` as an empty array. Check that you get a 400 and not a 500 response.
- [ ] Programmatically submit a short text field with no `answer`, but with `answerArray` as a non-empty array. Check that you get a 400 and not a 500 response.

### CODE REVIEW

File: src/app/utils/field-validation/field-validation.guards.ts
Comment (karrui) [inline src/app/utils/field-validation/field-validation.guards.ts:20]: nit: should the fieldType check come first? so the other checks are not for naught.

File: src/app/utils/field-validation/field-validation.guards.ts
Comment (karrui) [inline src/app/utils/field-validation/field-validation.guards.ts:37]: nit: should the fieldType check come first? so the other checks are not for naught.

File: src/app/utils/field-validation/field-validation.guards.ts
Comment (karrui) [inline src/app/utils/field-validation/field-validation.guards.ts:47]: nit: should the fieldType check come first? so the other checks are not for naught.

File: src/shared/resources/basic/index.ts
Comment (karrui) [inline src/shared/resources/basic/index.ts:7]: is this a backwards compatible change? the shared folder may not be updated on the clientside, meaning clients may not send over the `answerArray` key in the fields when they try to mutate a form and fail

**Review by karrui — APPROVED** [review]
lgtm, thanks for performing this refactor! looks good, just some client-side compatibility questions

File: src/shared/resources/basic/index.ts
Comment (mantariksh) [reply src/shared/resources/basic/index.ts:7]: this doesn't change what the clients send us, that's handled by the viewmodels and the `submitForm` function, which are independent of `shared`. the client's behaviour is completely unchanged because this `answerArray` field is only used on the server. so it doesn't matter if the client is updated or not

**Review by liangyuanruo — COMMENTED** [review]
@mantariksh could you take a look at #736? I believe there is some overlap in work.

**Comment by mantariksh** [comment]
@liangyuanruo yeah, I'll wait for #736 to be merged and then resolve merge conflicts

**Comment by mantariksh** [comment]
not to merge until release 4.48.1 is merged back into develop (will have merge conflicts)

**Comment by mantariksh** [comment]
@karrui @tshuli merge conflicts fixed, pls re-review due to extensive changes after merge conflicts with #736! thank you!

File: src/app/modules/spcp/spcp.util.ts
Comment (karrui) [inline src/app/modules/spcp/spcp.util.ts:190]: Should this just be ShortText also? It's not really an NRIC

File: src/app/routes/admin-forms.server.routes.js
Comment (karrui) [inline src/app/routes/admin-forms.server.routes.js:350]: use celebrate's  `[Segment.BODY]`?

File: src/app/routes/admin-forms.server.routes.js
Comment (karrui) [inline src/app/routes/admin-forms.server.routes.js:409]: samesies

**Review by karrui — APPROVED** [review]
lgtm

File: src/app/modules/spcp/spcp.util.ts
Comment (mantariksh) [reply src/app/modules/spcp/spcp.util.ts:190]: Isn't it an NRIC? I also did it for consistency with the decryption worker - see process-decrypted-content.js, L31

File: src/app/routes/admin-forms.server.routes.js
Comment (mantariksh) [reply src/app/routes/admin-forms.server.routes.js:350]: heh sorry missed that during copypasta

**Review by tshuli — APPROVED** [review]
lgtm

**Pull request merged**
